tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+//! stack.
+//!
+//! However it is sometimes desirable to preserve the exact ordering of events
+//! as they were emitted by `tracing-flame`, so you can see exactly when each
+//! span is entered relative to the other and get an accurate visual trace of
+//! the execution of your program. To get this style of representation you
+//! should instead generate a `flamechart`, which does not sort or collapse
+//! identical stack frames.
+//!
+//! [`tracing`]: https://docs.rs/tracing
+//! [`inferno`]: https://docs.rs/inferno
+//! [`FlameLayer`]: struct.FlameLayer.html
+mod error;
+
+use error::Error;
+use std::fmt;
+use std::fmt::Write as _;
+use std::fs::File;
+use std::io::BufWriter;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::path::Path;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    out: Mutex<W>,
+    last_event: Mutex<Instant>,
+    _inner: PhantomData<S>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    pub fn new(writer: W) -> Self {
+        Self {
+            out: Mutex::new(writer),
+            last_event: Mutex::new(Instant::now()),
+            _inner: PhantomData,
+        }
+    }
+}
+
+impl<S> FlameLayer<S, File>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    pub fn write_to_file(path: impl AsRef<Path>) -> Result<Self, Error> {
+        let path = path.as_ref();
+        let file = File::create(path).map_err(|source| Error::IO {
+            path: path.into(),
+            source,
+        })?;
+        // let writer = BufWriter::new(file);
+        Ok(Self::new(file))
+    }
+}
+
+impl<S, W> Layer<S> for FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    fn new_span(&self, _: &Attributes, id: &Id, ctx: Context<S>) {
+        let samples = self.time_since_last_event();
+
+        let mut spans = SpanIter::new(id.clone(), &ctx);
+        let _ = spans.next();
+        let spans = spans.rev();
+        let mut stack = String::new();
+
+        stack
+            .write_str("tracing")
+            .expect("expected: write to String never fails");
+
+        for parent in spans {
+            stack
+                .write_str("; ")
+                .expect("expected: write to String never fails");
+            write(&mut stack, parent).expect("expected: write to String never fails");
+        }
+
+        write!(&mut stack, " {}", samples.as_nanos())
+            .expect("expected: write to String never fails");
+        writeln!(
+            *self.out.lock().expect("expected: lock is never poisoned"),
```

Comment:
nit: this `expect` message kind of makes it sound like the lock being poisoned is an internal error/bug, which it isn't really...it's just a consequence of a panic elsewhere. i would _probably_ just unwrap here
```suggestion
            *self.out.lock().unwrap(),
```